pennylane/qnode_new/qnode.py

```diff
@@ -0,0 +1,920 @@
+        :class:`~pennylane.operation.Operator` instances (in this circuit) that depend on it.
+        """
+
+        self._metric_tensor_subcircuits = None  #: dict[tuple[int], dict[str, Any]]: circuit descriptions for computing the metric tensor
+        # introspect the quantum function signature
+        _get_signature(self.func)
+
+        self.output_conversion = None  #: callable: for transforming the output of :meth:`.Device.execute` to QNode output
+        self.output_dim = None  #: int: dimension of the QNode output vector
+
+
+    def __repr__(self):
+        """String representation."""
+        detail = "<QNode: device='{}', func={}, wires={}>"
+        return detail.format(self.device.short_name, self.func.__name__, self.num_wires)
+
+
+    def _set_variables(self, args, kwargs):
+        """Store the current values of the free parameters in the Variable class
+        so the Operators may access them.
+
+        Args:
+            args (tuple[Any]): positional (differentiable) arguments
+            kwargs (dict[str, Any]): auxiliary arguments
+        """
+        Variable.free_param_values = np.array(list(_flatten(args)))
+        if not self.mutable:
+            # only immutable circuits access auxiliary arguments through Variables
+            Variable.kwarg_values = {k: np.array(list(_flatten(v))) for k, v in kwargs.items()}
+
+
+    def _op_descendants(self, op, only):
+        """Descendants of the given operator in the quantum circuit.
+
+        Args:
+            op (Operator): operator in the quantum circuit
+            only (str, None): the type of descendants to return.
+
+                - ``'G'``: only return non-observables (default)
+                - ``'E'``: only return observables
+                - ``None``: return all descendants
+
+        Returns:
+            list[Operator]: descendants in a topological order
+        """
+        succ = self.circuit.descendants_in_order((op,))
+        if only == 'E':
+            return list(filter(_is_observable, succ))
+        if only == 'G':
+            return list(itertools.filterfalse(_is_observable, succ))
+        return succ
+
+
+    def _append_op(self, op):
+        """Append a quantum operation into the circuit queue.
+
+        Args:
+            op (~.operation.Operator): quantum operation to be added to the circuit
+        """
+        if op.num_wires == Wires.All:
+            if set(op.wires) != set(range(self.num_wires)):
+                raise ValueError("Operator {} must act on all wires".format(op.name))
+
+        # Make sure only existing wires are used.
+        for w in op.wires:
+            if w < 0 or w >= self.num_wires:
+                raise QuantumFunctionError("Operation {} applied to invalid wire {} "
+                                           "on device with {} wires.".format(op.name, w, self.num_wires))
+
+        # observables go to their own, temporary queue
+        if isinstance(op, Observable):
```

Comment:
Shouldn't you then use the predicate here? (Also in more places all over the file I didn't explicitly comment on)

Reply:
No, unfortunately. The queuing system is a bit complicated. The queuing of observables happens right after they are constructed, but they get assigned the `return_type` attribute only afterwards, via the measurement function (qml.expval etc). So, currently, they would not be recognized as "actual observables" here by the predicate.

Reply:
No, unfortunately it does not work. Observables are queued right after their construction, at which point they have a `return_type` of None, and thus enter the normal queue. Afterwards the measurement function (qml.expval etc) assigns them a return_type, removes them from the normal queue, and re-queues them.

This is a bit of a mess, but as long as we have classes like PauliX that can represent both observables and gates I can't see a perfectly clean solution.

Reply:
Me too, I think there is no better way of doing it, exactly because there are classes that are both Observable and Operation.